quic/s2n-quic-platform/src/syscall.rs

```diff
@@ -0,0 +1,176 @@
+    //# [RFC1191][RFC8201] for each flow utilizing DPLPMTUD and instead use
+    //# DPLPMTUD to control the size of packets that are sent by a flow.
+    #[cfg(s2n_quic_platform_mtu_disc)]
+    {
+        use std::os::unix::io::AsRawFd;
+
+        // IP_PMTUDISC_PROBE setting will set the DF (Don't Fragment) flag
+        // while also ignoring the Path MTU. This means packets will not
+        // be fragmented, and the EMSGSIZE error will not be returned for
+        // packets larger than the Path MTU according to the kernel.
+        success |= libc!(setsockopt(
+            tx_socket.as_raw_fd(),
+            libc::IPPROTO_IP,
+            libc::IP_MTU_DISCOVER,
+            &libc::IP_PMTUDISC_PROBE as *const _ as _,
+            core::mem::size_of_val(&libc::IP_PMTUDISC_PROBE) as _,
+        ))
+        .is_ok();
+
+        success |= libc!(setsockopt(
+            tx_socket.as_raw_fd(),
+            libc::IPPROTO_IPV6,
+            libc::IPV6_MTU_DISCOVER,
+            &libc::IP_PMTUDISC_PROBE as *const _ as _,
+            core::mem::size_of_val(&libc::IP_PMTUDISC_PROBE) as _,
+        ))
+        .is_ok();
+    }
+
+    success
+}
+
+/// Configures the socket to return TOS/ECN information as part of the ancillary data
+pub fn configure_tos(rx_socket: &Socket) -> bool {
+    let mut success = false;
+
+    #[cfg(s2n_quic_platform_tos)]
+    {
+        use std::os::unix::io::AsRawFd;
+        let enabled: libc::c_int = 1;
+
+        success |= libc!(setsockopt(
+            rx_socket.as_raw_fd(),
+            libc::IPPROTO_IP,
+            libc::IP_RECVTOS,
+            &enabled as *const _ as _,
+            core::mem::size_of_val(&enabled) as _,
+        ))
+        .is_ok();
+
+        success |= libc!(setsockopt(
+            rx_socket.as_raw_fd(),
+            libc::IPPROTO_IPV6,
+            libc::IPV6_RECVTCLASS,
+            &enabled as *const _ as _,
+            core::mem::size_of_val(&enabled) as _,
+        ))
+        .is_ok()
+    }
+
+    success
+}
+
+/// Configures the socket to return local address and interface information as part of the
+/// ancillary data
+pub fn configure_pktinfo(rx_socket: &Socket) -> bool {
+    let mut success = false;
+
+    // Set up the RX socket to pass information about the local address and interface
+    #[cfg(s2n_quic_platform_pktinfo)]
+    {
+        use std::os::unix::io::AsRawFd;
+        let enabled: libc::c_int = 1;
+
+        success |= libc!(setsockopt(
+            rx_socket.as_raw_fd(),
+            libc::IPPROTO_IP,
+            libc::IP_PKTINFO,
+            &enabled as *const _ as _,
+            core::mem::size_of_val(&enabled) as _,
+        ))
+        .is_ok();
+
+        success |= libc!(setsockopt(
+            rx_socket.as_raw_fd(),
+            libc::IPPROTO_IPV6,
+            libc::IPV6_RECVPKTINFO,
+            &enabled as *const _ as _,
+            core::mem::size_of_val(&enabled) as _,
+        ))
+        .is_ok();
```

Comment:
this previously had the check for ` if rx_addr.is_ipv4()`. Why did that get removed?

Reply:
It's not really needed to guard against. If the call fails because it's only a single IP-version stack, then that single call with fail. In this case, we consider the configuration successful if either/both of these calls succeed.